crates/task-impls/src/upgrade.rs

```diff
@@ -72,48 +72,34 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
     ) -> Option<HotShotTaskCompleted> {
         match event.as_ref() {
             HotShotEvent::UpgradeProposalRecv(proposal, sender) => {
-                let should_vote = self.should_vote;
+                info!("Received upgrade proposal: {:?}", proposal);
+
                 // If the proposal does not match our upgrade target, we immediately exit.
-                if !should_vote(proposal.data.upgrade_proposal.clone()) {
-                    warn!("Received unexpected upgrade proposal:\n{:?}", proposal.data);
+                if !(self.should_vote)(proposal.data.upgrade_proposal.clone()) {
```

Comment:
Is `(self.should_vote)` correct here? Does it return a function somehow?

Reply:
Yeah, `self.should_vote` has type `fn(UpgradeProposalData<TYPES>) -> bool`. There are probably a few ways of doing it; this just gives some flexibility in specifying exactly what proposals you want to accept, without having to change it directly in the consensus logic (you might want to vote yes on a specific upgrade but only within a certain range of view numbers, for example).

Reply:
Sounds good!